core/src/main/java/com/linecorp/armeria/common/logging/RequestLogAccess.java

```diff
@@ -0,0 +1,281 @@
+     *              if (headers.path().startsWith("/foo/")) {
+     *                  ...
+     *              }
+     *          });
+     * }</pre>
+     */
+    CompletableFuture<RequestLog> partialFuture(RequestLogProperty property);
+
+    /**
+     * Returns a {@link CompletableFuture} which will be completed when all the properties signified by
+     * the specified {@link RequestLogProperty}s are collected. The returned {@link CompletableFuture} is
+     * never completed exceptionally. Note that the completion of the returned {@link CompletableFuture}
+     * guarantees only the availability of the specified properties, which means any attempt to access
+     * other properties than specified may trigger a {@link RequestLogAvailabilityException}. If in doubt,
+     * use {@link #completeFuture()} or {@link #requestCompleteFuture()}.
+     * <pre>{@code
+     * logAccess.partialFuture(RequestLogProperty.REQUEST_HEADERS,
+     *                         RequestLogProperty.RESPONSE_HEADERS)
+     *          .thenAccept(log -> {
+     *              RequestHeaders reqHeaders = log.requestHeaders();
+     *              ResponseHeaders resHeaders = log.responseHeaders();
+     *              if (headers.path().startsWith("/foo/") &&
+     *                  resHeaders.status() == HttpStatus.OK) {
+     *                  ...
+     *              }
+     *          });
+     * }</pre>
+     *
+     * @throws IllegalArgumentException if {@code properties} is empty.
+     */
+    CompletableFuture<RequestLog> partialFuture(RequestLogProperty... properties);
+
+    /**
+     * Returns a {@link CompletableFuture} which will be completed when all the properties signified by
+     * the specified {@link RequestLogProperty}s are collected. The returned {@link CompletableFuture} is
+     * never completed exceptionally. Note that the completion of the returned {@link CompletableFuture}
+     * guarantees only the availability of the specified properties, which means any attempt to access
+     * other properties than specified may trigger a {@link RequestLogAvailabilityException}. If in doubt,
+     * use {@link #completeFuture()} or {@link #requestCompleteFuture()}.
+     * <pre>{@code
+     * logAccess.partialFuture(Lists.of(RequestLogProperty.REQUEST_HEADERS,
+     *                                  RequestLogProperty.RESPONSE_HEADERS))
+     *          .thenAccept(log -> {
+     *              RequestHeaders reqHeaders = log.requestHeaders();
+     *              ResponseHeaders resHeaders = log.responseHeaders();
+     *              if (headers.path().startsWith("/foo/") &&
+     *                  resHeaders.status() == HttpStatus.OK) {
+     *                  ...
+     *              }
+     *          });
+     * }</pre>
+     *
+     * @throws IllegalArgumentException if {@code properties} is empty.
+     */
+    CompletableFuture<RequestLog> partialFuture(Iterable<RequestLogProperty> properties);
```

Comment:
I'm not strong here, just summarizing what I was thinking while reviewing.

If users want to listen to all events and information of `RequestLogProperty.REQUEST_HEADERS`, `RequestLogProperty.REQUEST_HEADERS` and other properties immediately. And they want to make their own metric or logging.

I guess it is not impossible. It can be achieved by adding callback with event to `partialFuture(RequestLogProperty)` multiple times.
```java
logAcess.partialFuture(RequestLogProperty.REQUEST_START_TIME).thenAccept(log -> ...)
logAcess.partialFuture(RequestLogProperty.REQUEST_END_TIME).thenAccept(log -> ...)
logAcess.partialFuture(RequestLogProperty.RESPONSE_START_TIME).thenAccept(log -> ...)
logAcess.partialFuture(RequestLogProperty.RESPONSE_END_TIME).thenAccept(log -> ...)
```

Suggestion: How about adding `EventListener`, `ClientEventListener` which might need for DNS resolution timing information. Inspired by [okhttp - Events](https://square.github.io/okhttp/events/), [okhttp - EventListener](https://github.com/square/okhttp/blob/8511ee127f97f8cef2c382740b9cb6675d6ccae7/okhttp/src/main/java/okhttp3/EventListener.kt#L61)

```java
logAccess.addEventListener(new EventLister () {
	void onRequestHeaderSent(RequestHeaders) {
        ...
    }
    ...
})
```

Reply:
The API could be useful, but maybe we can add it later if there's a request for it. I think the use-case for grouped vs many partial listeners overwhelmingly leans towards grouped, in this PR there is basically only one location (on-child-request) that is setting many partial futures and in the wild it seems similarly rare.